### ISSUE TITLE
ArrayIndexOutOfBoundsException ZoneRules.isFixedOffset

### ISSUE THREAD

**Issue by looploooop** [Kotlin/kotlinx-datetime#149]
After updating to `kotlinx-datetime:0.3.0` I am seeing crashes on Android with the following stacktrace:

>java.lang.ArrayIndexOutOfBoundsException: length=0; index=-1
    at j$.time.zone.ZoneRules.isFixedOffset
    at kotlinx.datetime.TimeZone$Companion.ofZone$kotlinx_datetime(SourceFile:47)
    at kotlinx.datetime.TimeZone$Companion.currentSystemDefault(SourceFile:34)

I am not totally sure where to report this, the crash is introduced due to `0.3.0` making calls to `ZoneRules.isFixedOffset`.
Though by looking at the source code of `java.time.zone.ZoneRules` [here](https://github.com/google/desugar_jdk_libs/blob/master/src/share/classes/java/time/zone/ZoneRules.java#L506), I can't see how an `ArrayIndexOutOfBoundsException` can happen.

I have `coreLibraryDesugaring` enabled, using `com.android.tools:desugar_jdk_libs:1.1.5`

**Comment by dkhalanskyjb** [comment]
Could you please try calling `java.time.ZoneId.systemDefault()` and tell us what time zone is returned as a result?

**Comment by looploooop** [comment]
Currently I can see the two folloing timezones in sentry for users running into this crash:
> Africa/Cairo
America/Costa_Rica

I quickly changed my timezone to `America/Costa_Rica` and was able to reproduce the crash.

**Comment by dkhalanskyjb** [comment]
Are you able to reproduce this in the emulator? If so, could you please provide a reproducing project? I wasn't able to make the app crash in the emulator. Or is a device needed for the crash to occur?

Also, can you confirm that the line
```kotlin
java.time.ZoneId.of("America/Costa_Rica").rules.isFixedOffset
```
also crashes your app?

**Comment by looploooop** [comment]
>Are you able to reproduce this in the emulator? If so, could you please provide a reproducing project? I wasn't able to make the app crash in the emulator. Or is a device needed for the crash to occur?

It crashes in the emulator as well as on the device, [here](https://github.com/looploooop/java_zoneid_crash) is a small reproducer project. Basically I just copied your line of code into [here](https://github.com/looploooop/java_zoneid_crash/blob/main/app/src/main/java/com/example/datetimecrash/MainActivity.kt#L10) and enabled coreLibraryDeSugaring [here](https://github.com/looploooop/java_zoneid_crash/blob/main/app/build.gradle#L26)

>Also, can you confirm that the line...

Yes, I can confirm that the following line of code you posted is enough to produce the above mentioned crash.

**Comment by dkhalanskyjb** [comment]
Thanks! I managed to reproduce it on an emulator with the API level 24. With the API level 30, everything is fine.

I suppose it's really up to us to work around this to support older versions, as Google seems to have fixed this already in the newer SDKs.

**Comment by PaulWoitaschek** [comment]
Well this `$j` calls come from the desugar logic right? In addition to being worked around here this should be fixed upstream as well? 
That function doesn't exist in api 24 so it's a pure backport:
https://developer.android.com/reference/java/time/zone/ZoneRules#isFixedOffset()

@looploooop Didn't we have this issue on Android 30 as well?

Also what I don't understand here is where that crash is happening.

https://github.com/google/desugar_jdk_libs/blob/d1ca594532d3443822cc10c7a0856316e16f3e34/src/share/classes/java/time/zone/ZoneRules.java#L506-L517

https://github.com/google/desugar_jdk_libs/blob/d1ca594532d3443822cc10c7a0856316e16f3e34/jdk11/src/java.base/share/classes/java/time/zone/ZoneRules.java#L505-L516

https://github.com/google/desugar_jdk_libs/blob/d1ca594532d3443822cc10c7a0856316e16f3e34/jdk11/src/libcore/ojluni/src/main/java/java/time/zone/ZoneRules.java#L464-L466

All of these shouldn't trigger a `ArrayIndexOutOfBoundsException`

**Comment by looploooop** [comment]
>Didn't we have this issue on Android 30 as well?

Yes, looking at Sentry our users experienced this crash on API level versions 27-30

**Comment by dkhalanskyjb** [comment]
I've already prepared a workaround that should cover all the versions. Regarding reporting this to Google: were you able to reproduce this issue on API level 30?

**Comment by looploooop** [comment]
Yes, crashes on my Pixel 5 with API level 30 as well as in the Emulator with API level 31.

**Comment by dkhalanskyjb** [comment]
This is odd. Running your reproducing project in Android Studio 2021.1.1 on Linux, I get a crash on the  `Nexus 4 API 24` emulator, but no crash on `Pixel 2 API 30`, `Pixel 5 API 31`, or `Pixel 5 API 30`.

**Comment by looploooop** [comment]
Weird I just checked again, this crashes in an emulator `Pixel 5 API 31`, as well with Android Studio 2021.1.1 on Linux with` openjdk version "11.0.12" 2021-07-20
`
